ocaml/xapi/xapi_session.ml

```diff
@@ -27,6 +27,235 @@ open Client
 open Auth_signature
 open Extauth
 
+module AuthFail : sig
+  (* stats are reset each time you query, so if there hasn't
+     been a failed login attempt since the last time the stats
+     were queried, you won't get any stats *)
+  val get_stats_string : unit -> string option
+
+  val on_fail :
+       __context:Context.t
+    -> now:Date.iso8601
+    -> uname:string option
+    -> originator:string option
+    -> record:[< `log_only | `log_and_alert]
+    -> unit
+end = struct
+  type client = {
+      user_agent: string option
+    ; uname: string option
+    ; originator: string option
+    ; ip: string option
+  }
+
+  let client_of_info ~__context ~originator ~uname =
+    let user_agent = Context.get_user_agent __context in
+    let ip = Context.get_client_ip __context in
+    (* check to make sure we have at least _some_ information *)
+    if
+      [user_agent; originator; uname; ip]
+      |> List.for_all (function None | Some "" -> true | _ -> false)
+    then
+      None
+    else
+      Some {originator; uname; user_agent; ip}
+
+  let string_of_client x =
+    [
+      ("username", x.uname)
+    ; ("originator", x.originator)
+    ; ("useragent", x.user_agent)
+    ; ("ip", x.ip)
+    ]
+    |> List.filter_map (fun (label, value) ->
+           match value with
+           | None | Some "" ->
+               None
+           | Some value ->
+               Some (Printf.sprintf "<%s>%s</%s>" label value label))
+    |> String.concat "\n"
+
+  type client_failed_attempts = {
+      client: client
+    ; num_failed_attempts: int
+    ; last_failed_attempt: Date.iso8601
+  }
+
+  let up_to_3 xs x =
+    let rec merge = function
+      | [] ->
+          [x]
+      | y :: ys ->
+          if x.num_failed_attempts >= y.num_failed_attempts then
+            x :: y :: ys
+          else
+            y :: merge ys
+    in
+    let rec keep ctr xs =
```

Comment:
It would be enough to match the list; in Haskell this function is called `take`
```

let rec take n = function
| [] -> []
| x::xs when n > 0 -> x :: take (n-1) xs
| _ -> []
```

Reply:
This is available on Listext, there's a PR open for Stdlib as well.

Reply:
I've also replaced the merge function with a stable_sort

Reply:
Nice, thanks for tidying that up!

Reply:
Chaging code is easy when there are tests ;)